<a href="https://colab.research.google.com/github/rbryan2001/CSCI_591/blob/main/assignments/assignment_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 1 -  & Computational Programming

**Rebecca Bryan**

Netid: 790759512

Note: this assignment falls under collaboration Mode 2: Individual Assignment – Collaboration Permitted. Please refer to the syllabus on Canvas for additional information.

Instructions for all assignments can be found [here](https://github.com/kylebradbury/ids705/blob/master/assignments/_Assignment%20Instructions.ipynb), and is also linked to from the [course syllabus](https://kylebradbury.github.io/ids705/index.html).

Total points in the assignment add up to 90; an additional 10 points are allocated to presentation quality.

#  Learning Objectives
The purpose of this assignment is to provide a refresher on fundamental concepts that we will use throughout this course, and provide an opportunity to develop skills in any of the related skills that may be unfamiliar to you. Through the course of completing this assignment, you will...
- Practice numerical programming by loading and filtering data, plotting data, vectorizing operations, profiling code speed, and debugging and optimizing performance. You will also practice computing probabilities based on simulation.
- Develop or refresh your knowledge of Git version control, which will be a core tool used in the final project of this course
- Apply your skills altogether through an exploratory data analysis to practice data cleaning, data manipulation, interpretation, and communication

We will build on these concepts throughout the course, so use this assignment as a catalyst to deepen your knowledge and seek help with anything that is unfamiliar.


*Note: for all assignments, write out all equations and math using markdown and [LaTeX](https://tobi.oetiker.ch/lshort/lshort.pdf). For this assignment show ALL math work*

**ANSWER**

# Numerical Programming

## 1
**[30 points]** Loading data and gathering insights from a real dataset

In data science, we often need to have a sense of the idiosyncrasies of the data, how they relate to the questions we are trying to answer, and to use that information to help us to determine what approach we may need to apply to achieve our goal. This exercise provides practice in exploring a dataset and answering question that might arise from applications related to the data.


**Data**. The data for this problem can be found in the `data` subfolder in the `assignments` folder on [github](https://github.com/lucywowen/csci591_CCN). The filename is `stroke.csv`.

A stroke occurs when the blood flow to a part of the brain is reduced or restricted. Due to this brain cells start to die, in that part of the brain, at a very fast rate due to a lack of oxygen and nutrients. There are two types of brain strokes: (a) Ischemic stroke and (b) Haemorrhagic stroke of which ischemic stroke is more likely to occur. The rupture or blockage prevents blood and oxygen from reaching the brain’s tissues.
 Here we have used 8 input parameters like gender, age, various diseases, and smoking status in this dataset on brain stroke detection from Kaggle.
 The following information is provided about the patient:

    
|field    |description|
|:-----   |:-----|
|id       |unique identifier|
|gender   |'Male', 'Female', or 'Other'|
|age   |age of patient|
|hypertension    |0 if patient doesn't have hypertension; 1 if patient has hypertension|
|heart_disease      |0 if patient doesn't have heart disease; 1 if patient has heart disease|
|ever_married     |'No', 'Yes'|
|work_type |'children', 'Govt_jov', 'Never_worked', 'Private', 'Self-employed' |
|Residence_type   |'Rural', 'Urban' |
|avg_glucose_level |average glucose level of the patient|
|bmi |body mass index of the patient|
|smoking_status |'formerly smoked', 'never smoked', 'smokes', 'Unknown'|
|stroke |0 if patient has not had a stroke; 1 if patient has had a stroke|


**Your objective**. For this dataset, your goal is answer the following questions about these patients:

**(a)** Calculate the bmi mean and standard deviation for patients who have had a stroke. Calculate the bmi mean and standard deviation for patients who have had NOT had stroke.

**(b)** Assess the NA values in the data. Count the rows that have stroke as 0 and 1 respectively when bmi is null.

**(c)** Recalculate the bmi mean and standard deviation for patients who have had and have not had a stroke.  Plot this data as a box plot.

**(d)** Plot the same boxplot as above, but with the raw data plotted.  What additional information does this raw data provide? Briefly explain the differences between the plots shown in c and d.

**(e)** Plot a histogram of age distributions of the patients that have had strokes and patients that have not had strokes.  

**(f)** Plot a density plot of age distributions of the patients that have had strokes and patients that have not had strokes.  How is this plot different from the histogram plotted in e?  

**(g)** Plot a density plot of age distributions of the patients that have had strokes and patients that have not had strokes, by gender.  

**(h)** Plot a density plot of age distributions of the patients that have had strokes and patients that have not had strokes, by gender 'Male' and 'Female'. What changes with this plot relative to what you plotted for g?  

**ANSWER**

In [3]:
import pandas as pd
import numpy as np

In [1]:
!wget https://raw.githubusercontent.com/lucywowen/csci591_CCN/refs/heads/main/assignments/data/stroke.csv

--2025-01-30 21:34:46--  https://raw.githubusercontent.com/lucywowen/csci591_CCN/refs/heads/main/assignments/data/stroke.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 316971 (310K) [text/plain]
Saving to: ‘stroke.csv’

stroke.csv          100%[===================>] 309.54K  --.-KB/s    in 0.04s   

2025-01-30 21:34:47 (8.53 MB/s) - ‘stroke.csv’ saved [316971/316971]



In [5]:
stroke = pd.read_csv('stroke.csv')
stroke.head()

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1

In [41]:
stroke.shape

(5110, 12)

#### **(a)** Calculate the bmi mean and standard deviation for patients who have had a stroke. Calculate the bmi mean and standard deviation for patients who have had NOT had stroke.

In [6]:
stroke.groupby('stroke').bmi.agg(['mean','std'])

mean       std
stroke                     
0       28.823064  7.908287
1       30.471292  6.329452

Patients who have not had a stroke have a mean BMI of 28.82 with a standard deviation of 7.91. Patients who have had a stroke had a mean BMI of 30.47 with a standard deviation of 6.33.

#### (b) Assess the NA values in the data. Count the rows that have stroke as 0 and 1 respectively when bmi is null.

In [42]:
stroke.bmi.isna().sum()

201

In [39]:
nans = stroke[stroke.bmi.isna() == True]
nans.shape

(201, 12)

In [35]:
nans.stroke.value_counts()

stroke
0    161
1     40
Name: count, dtype: int64

Of the rows missing a BMI score, 161 had not had a stroke, and 40 had had a stroke.

#### **(c)** Recalculate the bmi mean and standard deviation for patients who have had and have not had a stroke. Plot this data as a box plot.

In [49]:
stroke1 = stroke.dropna(axis=0, how='any',subset=['bmi'])
grouped = stroke1.bmi.groupby(level = 'stroke')
grouped.boxplot(rot=45, fontsize=12, figsize=(8, 10))

ValueError: level name stroke is not the name of the index

# Version Control via Git

## 2
**[10 points]** Git is efficient for collaboration, and expectation in industry, and one of the best ways to share results in academia. You can even use some Git repositories (e.g. Github) as hosts for website, such as with the [course website](https://kylebradbury.github.io/ids705/index.html). As a data scientist with experience in machine learning, Git is expected. We will interact with Git repositories (a.k.a. repos) throughout this course, and your project will require the use of git repos for collaboration.

Complete the [Atlassian Git tutorial](https://www.atlassian.com/git/tutorials/what-is-version-control), specifically the following listed sections. Try each concept that's presented. For this tutorial, instead of using BitBucket as your remote repository host, you may use your preferred platform such as [Github](https://github.com/) or [Duke's Gitlab](https://gitlab.oit.duke.edu/users/sign_in).
1. [What is version control](https://www.atlassian.com/git/tutorials/what-is-version-control)
2. [What is Git](https://www.atlassian.com/git/tutorials/what-is-git)
3. [Install Git](https://www.atlassian.com/git/tutorials/install-git)
4. [Setting up a repository](https://www.atlassian.com/git/tutorials/install-git)
5. [Saving changes](https://www.atlassian.com/git/tutorials/saving-changes)
6. [Inspecting a repository](https://www.atlassian.com/git/tutorials/inspecting-a-repository)
7. [Undoing changes](https://www.atlassian.com/git/tutorials/undoing-changes)
8. [Rewriting history](https://www.atlassian.com/git/tutorials/rewriting-history)
9. [Syncing](https://www.atlassian.com/git/tutorials/syncing)
10. [Making a pull request](https://www.atlassian.com/git/tutorials/making-a-pull-request)
11. [Using branches](https://www.atlassian.com/git/tutorials/using-branches)
12. [Comparing workflows](https://www.atlassian.com/git/tutorials/comparing-workflows)

I also have created two videos on the topic to help you understand some of these concepts: [Git basics](https://www.youtube.com/watch?v=fBCwfoBr2ng) and a [step-by-step tutorial](https://www.youtube.com/watch?v=nH7qJHx-h5s).

For your answer, affirm that you *either* completed the tutorials above OR have previous experience with ALL of the concepts above. Confirm this by typing your name below and selecting the situation that applies from the two options in brackets.

**ANSWER**

*I, [**your name here**], affirm that I have [**completed the above tutorial / I have previous experience that covers all the content in this tutorial**]*

# Exploratory Data Analysis
## 3
**[50 points]** Here you'll bring together some of the individual skills that you demonstrated above and perform exploratory data analysis. Your goal is to explore the [datasets available](https://github.com/lucywowen/csci591_CCN/tree/main/projects) and identify questions or problems you're interested in working with.  Below, we walk through a process to follow for your analysis.

1. Find 3 datasets that interest you.  I provide you with several opensource here [here](https://github.com/lucywowen/csci591_CCN/tree/main/projects), but feel free to look for others!  
2. For each of the 3 datasets, describe the dataset, the source of the data, and the reason the dataset was of interest. What question are you hoping to answer through exploring the dataset?
3. Check the data and see if they need to be cleaned: are there missing values? Are there clearly erroneous values? Do two tables need to be merged together? Clean the data so it can be visualized. If the data are clean, state how you know they are clean (what did you check?).
3. Plot the data, demonstrating interesting features that you discover. Are there any relationships between variables that were surprising or patterns that emerged? Please exercise creativity and curiosity in your plots. You should have at least a ~3 plots exploring the data in different ways ... so at least 9 plots total.
4. What insights are you able to take away from exploring the data? Is there a reason why analyzing the dataset you chose is particularly interesting or important? Summarize this for a general audience - boil down your findings in a way that is accessible, but still accurate.
5. Define a clear question for each dataset that you would be interested in working on for the remainder of the semester.  Be prepared to pitch your project proposals to the class!


**ANSWER**